fairlearn/postprocessing/_tradeoff_curve_utilities.py

```diff
@@ -169,50 +175,47 @@ def _interpolate_curve(
     result : :class:`pandas:pandas.DataFrame`
         DataFrame with the points of the interpolated curve.
     """
-    data_transpose = data.transpose()
+    x_values = data[x_col].values
+    y_values = data[y_col].values
+    content_values = data[content_col].values
 
     content_col_0 = content_col + "0"
     content_col_1 = content_col + "1"
 
-    i = 0
-    dict_list = []
-    x0 = data_transpose[0][x_col]
-    while data_transpose[i + 1][x_col] == x0:
-        i += 1
-
-    # calculate the curve points for each x tick in x_grid
-    for x in x_grid:
-        # skip over data points that we've already passed
-        while x > data_transpose[i + 1][x_col]:
-            i += 1
-
-        # Calculate the y value at x based on the slope between data points i and i + 1
-        x_distance_from_next_data_point = data_transpose[i + 1][x_col] - x
-        x_distance_between_data_points = data_transpose[i + 1][x_col] - data_transpose[i][x_col]
-        p0 = x_distance_from_next_data_point / x_distance_between_data_points
-        p1 = 1 - p0
-        y = p0 * data_transpose[i][y_col] + p1 * data_transpose[i + 1][y_col]
-        dict_list.append(
-            {
-                x_col: x,
-                y_col: y,
-                P0_KEY: p0,
-                content_col_0: data_transpose[i][content_col],
-                P1_KEY: p1,
-                content_col_1: data_transpose[i + 1][content_col],
-            }
-        )
+    indices = np.searchsorted(x_values, x_grid, side="right") - 1
+    indices = np.clip(indices, 0, len(x_values) - 2)
+    indices[1:-1] = np.where(
```

Comment:
I only had to add this operation to not break the current tests. In principle the algorithm works without it, with a slightly different output that doesn't affect the behaviour of the `InterpolatedThresholder`.


Here's a detailed description of the change that break some tests (but does not affect the final results of the thresholding):

The difference only concerns cases where interpolation is not required:
- **Current Code (Main Branch), and this PR's code**:
  - If `x` (in `x_grid`) equals a value in `x_metric` at index `i`, the "interpolated" operations set `p0=0` (corresponding to the threshold at index i-1) and `p1=1` (corresponding to threshold at index i).
  - Exception: When `x` in `x_grid` coincides with `x_metric[0]` and `x_metric[1] > x_metric[0]`, it is the other way around.

- **This PR's code if we remove the `np.where`**:
  - If `x` (in `x_grid`) equals a value in `x_metric` at index `i`, the operations set `p0=1` (corresponding to threshold at index `i`) and `p1=0` (corresponding to the threshold at index `i+1`).
  - Exception: When `x` in `x_grid` coincides with the last element in `x_metric`, it is the other way around.

Notive how this does not affect the final output since the threshold with probability 1 remains unchanged, and the threshold with probability 0 does not matter. The only test expectations that would need to be updated are for test cases where there is no interpolation of operations/thresholds and where `x_metric[1] > x_metric[0]`.

Reply:
@adrinjalali what do you think about this ? Should I keep this `np.where` just to keep the test expectation strictly unchanged (more details in my comment above)?